src/cdk/tree/navigator.ts

```diff
@@ -0,0 +1,304 @@
+  }
+
+  /** Expand the current node if it's not expanded. Otherwise move to its first child. */
+  toggleCurrentFocusedNode() {
+    if (this.focusedData && this.treeControl) {
+      if (!this.treeControl.isExpanded(this.focusedData)) {
+        this.treeControl.expand(this.focusedData);
+      } else {
+        this.treeControl.collapse(this.focusedData);
+      }
+    }
+  }
+
+  /**
+   * Pass events to the keyboard manager. Available here for tests.
+   */
+  _handleKeydown(event: KeyboardEvent) {
+    let code = event.keyCode;
+
+    switch(event.keyCode) {
+      case HOME:
+        this.focusFirst();
+        break;
+      case END:
+        this.focusLast();
+        break;
+      case ENTER:
+        this.toggleCurrentFocusedNode();
+        break;
+      case UP_ARROW:
+        this.focusPreviousVisibleNode();
+        break;
+      case DOWN_ARROW:
+        this.focusNextVisibleNode();
+        break;
+      case LEFT_ARROW:
+        this.collapseCurrentFocusedNode();
+        break;
+      case RIGHT_ARROW:
+        this.expandCurrentFocusedNode();
+        break;
+      case NUMPAD_MULTIPLY_KEYCODE:
+        this.expandAll();
+        break;
+      case TAB:
+        this.tabOut.next();
+        return;
+    }
+    event.preventDefault();
+  }
+
+  _changeFocusedData(newFocused: T | undefined) {
+    if (newFocused) {
+      this.focusedData = newFocused;
+      if (this.nodeMap.get(this.focusedData)) {
+        this.nodeMap.get(this.focusedData)!.focus();
+      }
+    }
+  }
+
+  _getFirst(): T | undefined {
+    const nodeList = this._getNodeList();
+    return nodeList.length ? nodeList[0] : undefined;
+  }
+
+  _getLast(): T | undefined {
+    const nodeList = this._getNodeList();
+    return nodeList.length ? this._getLastChild(nodeList[nodeList.length - 1]) : undefined;
+  }
+
+  _getPrevious(): T | undefined {
+    if (!this.focusedData) {
+      return;
+    }
+    const parent = this.parentMap.get(this.focusedData);
+    const nodeList = this.nodeListMap.get(parent);
+    const index = nodeList!.indexOf(this.focusedData);
+    if (index === 0) {
+      return parent;
+    } else if (index > 0) {
+      return this._getLastChild(nodeList![index - 1]);
+    }
+  }
+
+  _getNext(): T | undefined {
+    if (!this.focusedData) {
+      return;
+    }
+    // Always return first child if the node is expanded
+    if (this.nodeListMap.has(this.focusedData) &&
+        this.treeControl && this.treeControl.isExpanded(this.focusedData)) {
+      const childNodeList = this._getNodeList(this.focusedData);
+      if (childNodeList.length) {
+        return childNodeList[0];
+      }
+    }
+    // Or return next sibling / parent's next child if any
+    let currentData: T | undefined = this.focusedData;
+    while(currentData) {
+      const parent = this.parentMap.get(currentData);
+      const nodeList = this.nodeListMap.get(parent);
+      const index = nodeList!.indexOf(currentData);
+      if (index === nodeList!.length - 1) {
```

Comment:
This is a bit of a stretch, but I think that the condition here could technically evaluate to true if you've got an node list. E.g. if the `nodeList.length` is 0, the `indexOf` will return -1 which is the same as `nodeList.length - 1`.

Reply:
If there's an empty nodeList, we should also go to parent, until we find a children list that have a next element.